File: app/src/main/java/tech/pauly/findapet/data/AnimalRepository.java
package tech.pauly.findapet.data;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.functions.Function;
import tech.pauly.findapet.BuildConfig;
import tech.pauly.findapet.data.models.AnimalListResponse;
import tech.pauly.findapet.data.models.FetchAnimalsRequest;
import tech.pauly.findapet.data.models.Filter;
import tech.pauly.findapet.data.models.StatusCode;

public class AnimalRepository {

    private static final int ANIMAL_RETURN_COUNT = 20;

    private AnimalService animalService;
    private ObservableHelper observableHelper;

    @Inject
    public AnimalRepository(AnimalService animalService, ObservableHelper observableHelper) {
        this.animalService = animalService;
        this.observableHelper = observableHelper;
    }

    public Observable<AnimalListResponse> fetchAnimals(FetchAnimalsRequest request) {
        Filter filter = request.getFilter();
        return animalService.fetchAnimals(request.getLocation(),
                                          BuildConfig.API_KEY,
                                          request.getAnimalType().getServerName(),
                                          request.getLastOffset(),
                                          ANIMAL_RETURN_COUNT,
                                          filter.getSex().getServerName(),
                                          filter.getAge().getServerName(),
                                          filter.getSize().getServerName(),
                                          filter.getBreed())
                            .flatMap(animalListResponse -> {
                                if (animalListResponse.getAnimalList() == null
                                    || animalListResponse.getAnimalList().size() == 0) {
                                    return Single.error(new PetfinderException(StatusCode.ERR_NO_ANIMALS));
                                }
                                return Single.just(animalListResponse);
                            })
                            .compose(observableHelper.applySingleSchedulers())
                            .toObservable();
    }
}

File: app/src/main/java/tech/pauly/findapet/data/AnimalRepository.kt
package tech.pauly.findapet.data

import io.reactivex.Observable
import tech.pauly.findapet.BuildConfig
import tech.pauly.findapet.data.models.AnimalListResponse
import tech.pauly.findapet.data.models.FetchAnimalsRequest
import tech.pauly.findapet.data.models.SingleAnimalResponseWrapper
import tech.pauly.findapet.data.models.StatusCode
import javax.inject.Inject

private const val ANIMAL_RETURN_COUNT = 20

open class AnimalRepository @Inject
constructor(private val animalService: AnimalService,
            private val observableHelper: ObservableHelper) {

    open fun fetchAnimals(request: FetchAnimalsRequest): Observable<AnimalListResponse> {
        val filter = request.filter
        return animalService.fetchAnimals(request.location,
                BuildConfig.API_KEY,
                request.animalType.serverName,
                request.lastOffset,
                ANIMAL_RETURN_COUNT,
                filter.sex.serverName,
                filter.age.serverName,
                filter.size.serverName,
                filter.breed)
                .map {
                    it.header.status?.code?.let {
                        if (it != StatusCode.PFAPI_OK) throw PetfinderException(it)
                    }
                    if (it.animalList == null || it.animalList?.size == 0) {
                        throw PetfinderException(StatusCode.ERR_NO_ANIMALS)
                    }
                    it
                }
                .compose(observableHelper.applySingleSchedulers())
                .toObservable()
    }

    open fun fetchAnimal(id: Int): Observable<SingleAnimalResponseWrapper> {
        return animalService.fetchAnimal(BuildConfig.API_KEY, id.toString())
                .map { SingleAnimalResponseWrapper(it, id) }
                .compose(observableHelper.applySingleSchedulers())
                .toObservable()
    }
}

